Server Consent API

xpple opened the pull request:
Dear FabricMC development team,

I hereby propose a server consent system for clients. As this is just a proposal, I invite you to share your thoughts on this so that we can come up with something better. Since this system is meant to be universal, I also opened an [issue over at Paper](https://github.com/PaperMC/Paper/issues/9380) for the (server-side) system. The text there is roughly the same, with a few server provider specific exceptions, and of course not including the client-sided aspect specifically for Fabric. Thanks for reading!

# Server Consent API

_Perhaps a different name would be better._

Summary
-------

This API adds a (universal) server consent system for mods and features
of mods a client connecting to the server uses. The server configures a list
of illegal mods/features and, if enabled, will send the client these lists
when the client connects to the server. Mod developers can then process
this information and if needed disable certain/all functionality. This system
is meant to be standardised across different server software
implementations, so that clients know what to implement.

Goals
-----

- Standardise a system for client mod developers wanting their mod to be
used responsibly. This is the main goal!
- Provide server owners with a system that can _ask_ the client to
disable some of its mods or some of its mods' functionalities.

Non-Goals
---------

It is absolutely not a goal to provide server owners with a robust
anti-cheat system. This system should not be used in place of anti-cheat
mods and other unfair gameplay preventing plugins/mods. This is not
the intention of this system.

Success Metrics
---------------

The amount of client-sided mods that will adopt this system. Though, it
should be noted that this system will not have failed if some mods don't
use this; for servers this system does not have any downsides and any mod
using this is therefore a success.

Motivation
----------

Client mod developers are often powerless to prevent their mods from
being used on servers which do not allow them/do not allow certain of
their functionalities. For instance, many cheat mod developers wouldn't
want their mods to be used outside of anarchy servers, but they have no
way of preventing it. Others might completely dismiss certain features
in fear that they will be used irresponsibly. With this system, client
mod developers are no longer limited in this regard, as they can now
disable their mod/illegal features on servers that want them to.

The server-side implementation as shown by the mod is very minimal,
and server owners only need to change a config to enable this system.
This is beneficial for server owners as it barely requires any effort to
configure. The same applies to the client-side implementation.

Description
-----------

This system uses the custom payload channel `noconsent:flags` to send
the list of mods/features to the client (namespace should be discussed).
The list is a `net.minecraft.util.Identifier` list with the following design:

- `<mod id>:<feature>` requests the feature `<feature>` of the mod with
mod id `<mod id>` to be disabled;
- `c:<feature>` requests the feature `<feature>` of all mods to be disabled;
- `<mod id>:all` requests the all the features of the mod with mod id
`<mod id>` to be disabled.

The common namespace `c` is defined by the field
`Flags#COMMON_NAMESPACE`. The wildcard feature `all` is defined by
the field `Flags#WILDCARD_FEATURE`. To aid standardisation, the `Flags`
class also includes common flags for mod developers to use.

If the server has enabled this system, the server will send the configured
list of mods/features to the client. For now, this list is configured using
JSON. The default config (located at `.minecraft/config/fabric/consents.json`
is as follows.

```json
{
  "enabled": false,
  "illegalFlags": []
}
```

An example config might look like this.

```json
{
  "enabled": true,
  "illegalFlags": [
    "c:markers",
    "seedmapper:all"
  ]
}
```

With this configuration, the server would ask all mods to disable their
`markers`, and to disable the mod with mod id `seedmapper` entirely. This
list is parsed by making a simple `TypeAdapter<Identifier>` provided by
GSON.

When the client receives the list of illegal mods/features, an event will
be triggered by this API to which client mod developers can listen. The
events is `ClientFabricServerConsentFlagsCallback.FLAGS_SENT`. They can
then check if their mod is included in any way in the list by using the
`ClientFabricServerConsent#isIllegal` method:

```java
/**
 * Checks for a given flag and mod id whether the flag is illegal.
 *
 * @param flag the flag to check against
 * @return {@code true} if the flag is illegal, {@code false} otherwise
 */
public static boolean isIllegal(Identifier flag) {
	for (Identifier illegalFlag : ClientFabricServerConsentImpl.illegalFlags) {
		if (illegalFlag.getNamespace().equals(Flags.COMMON_NAMESPACE)) {
			if (illegalFlag.getPath().equals(flag.getPath())) {
				return true;
			}
		}

		if (illegalFlag.getNamespace().equals(flag.getNamespace())) {
			if (illegalFlag.getPath().equals(Flags.WILDCARD_FEATURE)) {
				return true;
			}

			if (illegalFlag.getPath().equals(flag.getPath())) {
				return true;
			}
		}
	}

	return false;
}
```

It is up to the mod developer to decide what to do with this
information. 

Alternatives
------------

- The system is standardised by server software providers, but not
implemented into them directly. This means that different server
software providers would agree on a namespace for the list of
illegal mods/features, but let the configuration and implementation
be up to server owners to implement. One way this could be
achieved is by having server software providers document the
system clearly, with example code for server owners. There are two
downsides of this approach, though.

  1. The documentation will likely not reach the same audience as
  direct implementation would. That means fewer server owners
  will implement use system, which weakens the system as whole.
  2. It is at the expense of standardisation. Direct implementation
  simply has more of a standardising effect than individual
  standardisation has.

- This system is not standardised at all. In that case a mod (in the
case of a Fabric server) could be used by server owners. However,
there is very little chance a third party mod will be widely adopted
by servers. Importantly, standardisation is the main goal of this
system. Without it, disagreement and inconsistency would arise, and
the system would flop. Only if established server software providers
choose to adopt this system will this work.

Considering these two alternatives, only the first could actually work.
Third party implementations just don't have the necessary reach to
establish something widespread across the Minecraft developer
community.

xpple committed:
Foundation for permissions api

xpple committed:
Foundation for server consent api

xpple committed:
Resolve style check

xpple committed:
Resolve style check 2

xpple commented:
It has become clear that the client sending a list of their mods to the server is a bad idea. I will rewrite the code and PR description accordingly later.

xpple committed:
Reimplement Server Consent API

xpple committed:
Minor style changes

xpple committed:
Fix `enabled` config being ignored + make channel more universal

modmuss50 commented on fabric-server-consent-api-v1/src/client/java/net/fabricmc/fabric/api/server/consent/v1/client/ClientFabricServerConsentEvents.java:
```diff
@@ -0,0 +1,55 @@
+import net.fabricmc.fabric.api.networking.v1.PacketSender;
+
+public final class ClientFabricServerConsentEvents {
+	/**
+	 * An event to notify the client that the server has sent their list of illegal mods.
+	 */
+	public static final Event<ModsSent> MODS_SENT = EventFactory.createArrayBacked(ModsSent.class, callbacks -> (client, handler, buf, responseSender) -> {
+		for (ModsSent callback : callbacks) {
+			callback.onModsSent(client, handler, buf, responseSender);
+		}
+	});
```
Remove everything specific to mods

modmuss50 commented on fabric-server-consent-api-v1/src/client/java/net/fabricmc/fabric/api/server/consent/v1/client/ClientFabricServerConsentEvents.java:
```diff
@@ -0,0 +1,55 @@
+
+	@FunctionalInterface
+	public interface FeaturesSent {
+		void onFeaturesSent(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf, PacketSender responseSender);
```
I have no clue what this event is trying to do?

xpple replied:
We want to make it as easy as possible for clients to adopt this system. Mod developers can use this event to possibly disable functionality when the server asks them to. Along with this I added a field which holds the list of illegal flags as well, so the client will always be able to access it.

modmuss50 commented on fabric-server-consent-api-v1/src/main/java/net/fabricmc/fabric/api/server/consent/v1/FabricServerConsent.java:
```diff
@@ -0,0 +1,51 @@
+	 *
+	 * @return the list of illegal features
+	 */
+	public static List<String> getIllegalFeatures() {
```
`public static List<Identifier> getFlags()`

I dont think features is quite right, flags might be better?

maityyy replied:
Maybe "mod flags"? Since 1.19.4 we have "feature flags"

xpple replied:
I have changed it to `flags` for now.

modmuss50 reviewed (changes requested):
This is going to need a lot of work before its ready, I have left a very high level review based on discussion.

xpple committed:
Implement requested changes

xpple committed:
Execute code on the correct thread

xpple committed:
Mark API as experimental

xpple committed:
Add example commands + rewrite isIllegal method

xpple committed:
Text.literal -> Text.of

modmuss50 committed:
Make some experimental APIs stable. (#3354)

* Make some experimental APIs stable.

* Imports.

LLytho committed:
Add LootTableEvents.LOADED event  (#3352)

* Implement `LootTableEvents.LOADED` event

* Update for checkstyle

* rename event

* Update fabric-loot-api-v2/src/main/java/net/fabricmc/fabric/api/loot/v2/LootTableEvents.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

---------

Co-authored-by: modmuss <modmuss50@gmail.com>
Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>
(cherry picked from commit 96dfa9590d310c7758c57c41c713d6176ae69f7f)

apple502j committed:
Fix inconsistency of placed feature locations (#3369)

* Fix inconsistency of placed feature locations

`BiomeSource#getBiomes` mixin applies to all biome sources, including one for Overworld.
The return value is a set; however one caller in the worldgen code iterates over it: `PlacedFeatureIndexer`.
Using a hash set here randomizes the return value, affecting feature placement.
Use a linked hash set instead.

* Improve fix to only make changes when required.

---------

Co-authored-by: modmuss50 <modmuss50@gmail.com>

modmuss50 committed:
Bump version

AnOpenSauceDev commented:
I like the idea of this, but I doubt that this will have a major effect in the end, at least on the side of hacked clients where someone could easily just maintain a fork that ignores the API. After all, most major hacked clients don't even bother to check if the server is anarchy or not. In my opinion it's just too much work for something that probably wont be used often or can easily be bypassed.

xpple commented:
> I like the idea of this, but I doubt that this will have a major effect in the end, at least on the side of hacked clients where someone could easily just maintain a fork that ignores the API. After all, most major hacked clients don't even bother to check if the server is anarchy or not. In my opinion it's just too much work for something that probably wont be used often or can easily be bypassed.

It's up to the developer to adopt this system or not. And yes, it can be easily bypassed. It's an explicit Non-Goal to provide some robust anti-cheat system. This is for client mod developers that want their mod to be used more responsibly.

gudenau commented:
This would be trivial to bypass as the stated goal is to disabled cheats on servers any "client" could just disable this check. Even mods that use this could have the checks disabled.

AnOpenSauceDev commented:
> This would be trivial to bypass as the stated goal is to disabled cheats on servers any "client" could just disable this check. Even mods that use this could have the checks disabled.

I will say, it would totally be possible for someone to make a mod* that consists of a few mixins, that bypasses the check for all mods, which would mean that compliant mod devs would still have their mods misused. It kind of sucks that there isn't a good way to do with without a) being some level of invasive or b) obfuscating code.

>*modrinth does not allow these forms of mods, but GitHub and Curseforge could possibly distribute this.

xpple committed:
Merge remote-tracking branch 'upstream/1.20.2' into permissions-api

# Conflicts:
#	gradle.properties
#	settings.gradle

xpple committed:
Send flags during configuration phase

xpple committed:
Update build.gradle

modmuss50 committed:
Test fix: Object Builder API's Testmod crashing on windows. Closes #3480

kevinthegreat1 committed:
Add Client After Block Break Event (#3367)

* Add clientside after block break event

* Update tests

* Checkstyle, of course

* Update Javadoc

* New Event

* Checkstyle 2 electric boogaloo

* Remove block entity parameter

* Refactor ClientPlayerBlockBreakEvents

* Update Javadoc

modmuss50 committed:
Bump version

modmuss50 committed:
Fix race condition when handling unknown packets. (#3508)

modmuss50 committed:
Bump version

deirn committed:
fix jukebox state getting changed mid-transaction (#3517)

SquidDev committed:
Check for the correct BE type in the lookup API (#3510)

The block lookup API's registerForBlockEntities method currently just
registers the passed provider for each valid block for that block
entity.

However, in some situations (such as update suppression or a misbehaving
mod), the wrong block entity will be present for one of these blocks.
This means that the (incorrect) block entity will be passed off to the
provider. Providers (especially those registered with
registerForBlockEntity) will often blind-cast the supplied BE, leading
to crashes.

While the wrong BE being present is a bug, we should follow vanilla's
lead and handle this more gracefully. In this case, we just check
whether the correct BE type is present before forwarding it to the
main provider.

ErrorCraft committed:
Add a system property for a custom output directory for gametest structures (#3509)

* Add a check for the output directory in the test command

* Fix style and add a run configuration with the output directory set

* Update fabric-gametest-api-v1/build.gradle

---------

Co-authored-by: modmuss <modmuss50@gmail.com>

deirn committed:
fix local login query response handler not working (#3495)

jpenilla committed:
add fabric-api-bom and fabric-api-catalog (#3487)

* feat: add fabric-api-bom and fabric-api-catalog

* spotlessApply

* Apply suggestions from code review

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

---------

Co-authored-by: modmuss <modmuss50@gmail.com>
Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Bump version

modmuss50 committed:
Exclude meta projects from main maven artifact.

This fixes being unable to depend on it in dev.

modmuss50 committed:
Bump version

Syst3ms committed:
Add mob conversion event (#3511)

* Add mob conversion event

* spacing

* use a smarter mixin strategy

* add pig and villager lightning triggers

* tyop

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

* handle tadpole -> frog conversion

- update javadoc to clarify the handled cases
- document keepEquipment parameter
- improve mixin handler names

---------

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

modmuss50 committed:
Stack aware `getFoodComponent` (#3520)

* Stack aware `getFoodComponent`

Co-authored-by: Phoupraw <50520903+phoupraw@users.noreply.github.com>

* Back to redirects

---------

Co-authored-by: Phoupraw <50520903+phoupraw@users.noreply.github.com>

jaredlll08 committed:
Don't tick gametests if the server isn't ticking (#3528)

* Don't tick gametests if the server isn't ticking

* Fix import order

modmuss50 committed:
Update to loom 1.5 (#3515)

* Update to loom 1.5

* Continue to use the mixin AP for now.

* Bump

(cherry picked from commit 05ec3fa1a913df35e518cdfd1d37814ef66ea4af)

modmuss50 committed:
Bump version

deirn committed:
fix login network addon not properly initialized (#3531)

modmuss50 committed:
Bump version

Syst3ms committed:
Data Attachment API (#3476)

* Data Attachment API

* javadoc

* Remove AttachmentSerializer & independent syncability and persistence

- removed AttachmentSerializer in favor of codecs
- renamed serializability to "persistence"
- made persistence and syncability independent switches
- reworked convenience registry methods to use Suppliers from the get-go

* Move some serialization-related methods to impl

- changed logger name

* rename Attachment to AttachmentType

* Added DefaultedAttachmentType and reworded Javadoc

* add warning in getAttached

* javadoc

* fix defaulted API

* Add unit tests

* remove DefaultedAttachmentType, add helper methods

bikeshedding inbound

* add more unit tests

* add testmod

it works

* stash syncing for a further PR

* missed license header

* address most reviews

* more reviews

* naming convention

* fix tyop

* fix invalid file name error

* simplify API in the absence of sync

It was established that the presence of a codec might be useful for other things than persistence, and while this seems to couple the two, the API can be changed later in a backward-compatible way if need be.

* couple codec with persistence

committing to the change I mentioned previously

* little fixes

* Fix mixins + requests

- Copy attachments on entity copy,& with a customizable flag in the case of player respawn
- Call relevant change-notifying methods when calling setAttached on BEs and chunks
- Change persistence key
- Fix mixin visibility

* Write tests for entity copy

* replace mixin by COPY_FROM event

* missed license header

* more advanced copy mechanics

- attachments require an EntityCopyHandler to be copied across entities
- a copy handler is automatically derived if there's a codec
- updated javadoc for chunk and BE targets

* Revert "more advanced copy mechanics"

This reverts commit 3f53b554fbd1c64fd21c95c067d14d43b9b77063.

* replaced complicated API by a stern warning

- also handled cross-world entity teleportation

* add gametest

* fix compilation

* flipped boolean

* forgot some more bools to flip

* requests

* fix FMJ

* fix BE mixin and add gametest

* add client player entity copying

* Use new mob conversion event

---------

Co-authored-by: modmuss <modmuss50@gmail.com>

PepperCode1 committed:
Add AtlasSourceTypeRegistry (#3504)

modmuss50 committed:
Bump version

deirn committed:
send `c:register` packet for play phase instead of config (#3544)

modmuss50 committed:
Bump version

apple502j committed:
Fix TAGS_LOADED event not being invoked on client join

apple502j committed:
Refactor resource loader internals (#3473)

* First step toward fixing resource pack grouping

* Placeholder pack and pack dependency

* Various fixes

* Fix wrong variable in serialization code

* Hide packs in PackScreen and DatapackCommand

* Apparently Japanese people aren't alone in having their currency signs used as special chars...

* Inject directly to Pack

* Add temporary logging, fix bug

* Add proper sorting

* Improve logging

* Fix duplicate name registration

* Fix client pack handling

* Fix FMJ

* Stop using interface injection for internal interface

* Delete unused GroupResourcePack

* Move refreshAutoEnabledPacks to util

* Improve logging

* Make a few things private

* Use vanilla metadata serialization logic

* Improve javadoc

* Add junit test

* Some final refactors

* Update ja_jp.json

---------

Co-authored-by: modmuss <modmuss50@gmail.com>

ErrorCraft committed:
Allow `FabricCodecDataProvider` to access dynamic registries (#3522)

* Add CompletableFuture to FabricCodecDataProvider

* Deprecate the old method and constructor, and fix the style

* Fix the style for real this time

* Add exceptions to the configure methods

* Apply suggestions from code review

---------

Co-authored-by: modmuss <modmuss50@gmail.com>

TelepathicGrunt committed:
Make chunk sections only convert vanilla air blocks to AIR (#3535)

* Make chunk sections only convert vanilla air blocks to AIR

* angry checkstyles calmed

* Comments added for future reference

modmuss50 committed:
Bump version

apple502j committed:
Fix duplicate resource pack names (#3557)

modmuss50 committed:
Bump version

jellysquid3 committed:
Expose a function for querying the non-default fluid renderer (#3558)

* Expose a function for querying the non-default fluid renderer

* Do not register default handlers as overrides

* Add @Nullable to getOverride

* Add @Nullable to get

---------

Co-authored-by: modmuss50 <modmuss50@gmail.com>

jacobsjo committed:
Extend Data Attachment API to ProtoChunk (#3548)

* allow data-attachment on ProtoChunks

- moved interfaceInjection from WorldChunk to Chunk
- dataAttachment saving on ProtoChunks in ChunkSerializer
- copy attachment from ProtoChunk to WorldChunk on creation.
- make WrapperProtoChunk wrap attachment calls to WorldChunk

* add test for data-attachment on ProtoChunks, and extend testmod.

* code style and license headers

* fix typos in javadoc

* extend testmod to test setting attachment during worldgen.

* code formatting

* fix testmod: don't crash when feature isn't placed (i.e. on GameTest server)

* add warning when adding persistent attachment to chunk with status EMPTY.

* update javadoc

* update javadoc to reference ServerLivingEntityEvents#MOB_CONVERSION

MrNavaStar committed:
Add Save Events to Server Lifecycle Events (#3533)

* Add Server Begin Save and Server End Save event

* rename events

* add tests

* pass more params through

PepperCode1 committed:
Add Support for Custom ColorResolvers (#3503)

* Add support for custom color resolvers

* Add ColorResolverRegistry

* Fix checkstyle

* Statically initialize all BiomeColorCaches

modmuss50 committed:
Bump version

modmuss50 committed:
Fix coverage task (#3583)

(cherry picked from commit 3335cde7ee40cc2760b6b69bda141e621cab06e1)

jacobsjo committed:
Fix data attachment testmod (#3586)

* fix data attachment testmod: use marker file to determine first/second launch

* fix typos

* one more typo

modmuss50 committed:
Fix empty data attachments being saved. (#3588)

* Fix empty data attachments being saved.

* Checkstyle

modmuss50 committed:
Bump version

PepperCode1 committed:
Fluid rendering fixes (#3593)

* Fluid rendering fixes

- Fix default overlay block check using TransparentBlock instead of TranslucentBlock
- Fix setBlockTransparency/isBlockTransparent not being thread-safe
- Reuse the same render handler objects for water and lava

* Fix implementation issues

- Fix custom geometry being buffered twice if FluidRenderHandler#renderFluid is invoked directly
- Fix calling FluidRenderHandler.super.renderFluid not using passed arguments to calculate color
- Fix calling FluidRenderHandler.super.renderFluid more than once producing incorrect geometry
- Fix fluids with no handler never receiving water overlay instead of using default behavior

* Add way to render fluid with non-vanilla default

- Fix testmod

Syst3ms committed:
Remove unnecessary recreation of identifiers in VillagerTypeHelper.register (#3597)

Syst3ms committed:
Improve handling of empty item groups in the creative menu (#3600)

* Better handling of empty item groups in pagination

Empty/non-displayable item groups are now put after all other groups and only then are they split into pages.

* Serendipitious bugfix

Fixed an issue where empty item groups could get selected when switching pages

* checkstyle

modmuss50 committed:
Fix teal sign testmod crash by registering the wood type. (#3599)

(cherry picked from commit 98c5af8bcb97dc59821a979ee0307e7e8dc765b0)

modmuss50 committed:
Bump version

modmuss50 committed:
Setup crowdin (#3604)

FabricMCBot committed:
Translastion updates (#3605)

* New translations en_us.json (Spanish)

* New translations en_us.json (Bulgarian)

* New translations en_us.json (German)

* New translations en_us.json (Greek)

* New translations en_us.json (Finnish)

* New translations en_us.json (Japanese)

* New translations en_us.json (Korean)

* New translations en_us.json (Polish)

* New translations en_us.json (Swedish)

* New translations en_us.json (Turkish)

* New translations en_us.json (Chinese Simplified)

* New translations en_us.json (Chinese Traditional)

* New translations en_us.json (Vietnamese)

* New translations en_us.json (Portuguese, Brazilian)

* New translations en_us.json (Spanish, Mexico)

* New translations en_us.json (Estonian)

* New translations en_us.json (Tatar)

* New translations en_us.json (French)

* New translations en_us.json (Spanish)

* New translations en_us.json (Polish)

* New translations en_us.json (Russian)

* New translations en_us.json (Ukrainian)

* New translations en_us.json (Chinese Traditional)

* New translations en_us.json (Estonian)

* New translations en_us.json (Russian)

* New translations en_us.json (French)

* New translations en_us.json (Spanish)

* New translations en_us.json (Polish)

* New translations en_us.json (Russian)

* New translations en_us.json (Ukrainian)

* New translations en_us.json (Estonian)

* New translations en_us.json (Dutch)

* New translations en_us.json (Dutch)

* New translations en_us.json (Dutch)

* New translations en_us.json (Finnish)

* New translations en_us.json (Finnish)

* New translations en_us.json (Italian)

* New translations en_us.json (Italian)

* New translations en_us.json (Swedish)

* New translations en_us.json (German)

* New translations en_us.json (German)

* New translations en_us.json (German)

* New translations en_us.json (Chinese Simplified)

* New translations en_us.json (German)

maityyy committed:
Fire the `MOB_CONVERSION` event after shearing the mushroom cow (#3612)

* Fire the MOB_CONVERSION event after shearing the mushroom cow

* checkstyle

* Add license header

modmuss50 committed:
Bump version

xpple committed:
Merge remote-tracking branch 'upstream/1.20.4' into permissions-api

# Conflicts:
#	gradle.properties

xpple committed:
Trigger check

xpple committed:
Trigger check

xpple committed:
Fix channel receiver registration bug

TelepathicGrunt committed:
Default English Translations for `c` namespaced tags (#3716)

* Created tag translation datagen

* generate lang file

* Add fruit translation

* renamed interface injection

* Adjust missing translation message

* Fixed for checkStyle

* More checkstyle fighting...

* Bump version of modules.

* Oops wrong module got bumped

* moving and renaming files

* adjust enum class naming

* rename enum and move package

* Renamed one more enum

* Undo version bump. Will be handled at release

* Some suggestions implemented

* Added crowdin entry

* Update crowdin.yml

Co-authored-by: modmuss <modmuss50@gmail.com>

* Cleanup enum checking

* spotless fixes

* Added buckets and powder snow bucket tag translation

* Some more suggested changes

* setup ConventionLogWarnings to match ConventionLogWarningsClient setup

* Change property to be usable in prod

* Set short as default property

* Remove unused import

* Fixed copy/paste error

* fixed ingot translations

* lang and method name fixes

* Fixed filepath issue

* Fixed redstone name

* fixed javadoc whitespace

* Move translation warning to server

* Move translation class to better file path

* trying to pass the regex

* put it under impl

* Cleanup translation check

* Update fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricLanguageProvider.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

* Update fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricLanguageProvider.java

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

* Update fabric-convention-tags-v2/src/main/java/net/fabricmc/fabric/api/tag/FabricTagKey.java

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

* Update fabric-convention-tags-v2/src/main/java/net/fabricmc/fabric/api/tag/FabricTagKey.java

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

---------

Co-authored-by: modmuss <modmuss50@gmail.com>
Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>
Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

jacobsjo committed:
Use knownPackInfo on mod data packs to avoid registry syncing (#3726)

* use knownPack to avoid syncing mod data packs.

* remove empty class

* checkstyle and licenses

* move debug logging from testmod into main module

* fix knownPackInfo not working for build-in data packs.

* make bundled data packs not required for client, and don't auto enable

* improve testmod with added builtin data pack + better error message

* Increase max known packs in C2S packet

* validate size before sending
* changeable with system property

* change mixin from ModifyConstant to ModifyArg to be more clear.

* Apply suggestions from code review

Co-authored-by: modmuss <modmuss50@gmail.com>

* store list of knownPacks as server start

avoids desync on pack change (since registries aren't reloaded)

* be extra safe: only request knownPacks that were enabled at server start and still are.

Doesn't rely on the fact that registries aren't synced anymore.

---------

Co-authored-by: modmuss <modmuss50@gmail.com>

modmuss50 committed:
Add API to modify default item components (#3728)

* Add API to modify default item components

* Add test for removal

* Some review feedback

* API design changes

* Review feedback

* Add overload that takes a Collection<Item>

modmuss50 committed:
Don't enforce a RegistryByteBuf packet code for the extended screen handler type (#3743)

SquidDev committed:
Support extending dynamic registries in datagen (#3744)

- Add all dynamic registries to the initial registry.
 - Expose a getter for the registry future.
 - Register cloners for all dynamic registries.

Apollounknowndev committed:
Support loading a single resource condition instead of array (#3749)

* Support loading a single condition instead of array

* Remove extra space

* Use LIST_CODEC field instead of CODEC.listOf()

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

---------

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
1.20.6

modmuss50 committed:
Bump version

apple502j committed:
Fix non-player handling in CustomDamageHandler (#3781)

(cherry picked from commit 5b61bd2784cc453836ecd18ca166324760b8075e)

kevinthegreat1 committed:
Fix ClientMessageEvents Injection Points (#3778)

SquidDev committed:
Registry resource conditions (#3752)

* Add resource conditions to dynamic registries

* Support conditions in FabricDynamicRegistryProvider

 - Add matching overloads for all add(...) methods which take a varargs
   list of ResourceConditions.

 - Add an additional overload for registering RegistryEntry.References
   directly. This makes it a little easier to register directly from a
   RegistryBuilder.

* Throw error if we cannot add resource conditions

* Rename EntryWithConditions to ConditionalEntry

modmuss50 committed:
Bump version

PepperCode1 committed:
Fix model offset being ignored in non-terrain rendering (#3799)

- Add exception handling to BlockRenderContext#render to match vanilla

opekope2 committed:
Fix FabricItem.getAttributeModifiers (#3797)

modmuss50 committed:
Suppress translated tag warnings in prod (#3791)

(cherry picked from commit ea7ee968d4e9e95d78dac6181bb0e9a15c30b7ab)

modmuss50 committed:
Bump version

modmuss50 committed:
Revert fix for FabricItem.getAttributeModifiers() and deprecate for removal.

Closes #3812

xpple committed:
Merge remote-tracking branch 'refs/remotes/upstream/1.20.6' into permissions-api

# Conflicts:
#	build.gradle
#	crowdin.yml
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentTarget.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/impl/attachment/AttachmentPersistentState.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/impl/attachment/AttachmentSerializingImpl.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/impl/attachment/AttachmentTargetImpl.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/mixin/attachment/AttachmentTargetsMixin.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/mixin/attachment/BlockEntityMixin.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/mixin/attachment/ChunkSerializerMixin.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/mixin/attachment/EntityMixin.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/mixin/attachment/ServerWorldMixin.java
#	fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/mixin/attachment/WrapperProtoChunkMixin.java
#	fabric-data-attachment-api-v1/src/main/resources/fabric-data-attachment-api-v1.mixins.json
#	fabric-data-attachment-api-v1/src/main/resources/fabric.mod.json
#	fabric-data-attachment-api-v1/src/test/java/net/fabricmc/fabric/test/attachment/CommonAttachmentTests.java
#	fabric-data-attachment-api-v1/src/testmod/java/net/fabricmc/fabric/test/attachment/gametest/BlockEntityTests.java
#	fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricCodecDataProvider.java
#	fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestEntrypoint.java
#	fabric-events-interaction-v0/src/client/java/net/fabricmc/fabric/api/event/client/player/ClientPlayerBlockBreakEvents.java
#	fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java
#	fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItemStack.java
#	fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java
#	fabric-item-api-v1/src/main/resources/fabric-item-api-v1.mixins.json
#	fabric-item-api-v1/src/testmod/resources/fabric.mod.json
#	fabric-item-group-api-v1/src/main/resources/assets/fabric/lang/ko_kr.json
#	fabric-lifecycle-events-v1/src/client/java/net/fabricmc/fabric/mixin/event/lifecycle/client/ClientPlayNetworkHandlerMixin.java
#	fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/mixin/networking/client/ClientCommonNetworkHandlerMixin.java
#	fabric-networking-api-v1/src/testmod/java/net/fabricmc/fabric/test/networking/play/NetworkingPlayPacketTest.java
#	fabric-networking-api-v1/src/testmodClient/java/net/fabricmc/fabric/test/networking/client/play/NetworkingPlayPacketClientTest.java
#	fabric-registry-sync-v0/src/main/resources/assets/fabric-registry-sync-v0/lang/de_de.json
#	fabric-registry-sync-v0/src/main/resources/assets/fabric-registry-sync-v0/lang/et_ee.json
#	fabric-registry-sync-v0/src/main/resources/assets/fabric-registry-sync-v0/lang/fr_fr.json
#	fabric-rendering-v1/src/testmodClient/java/net/fabricmc/fabric/test/rendering/client/CustomAtlasSourcesTest.java
#	fabric-resource-loader-v0/src/client/java/net/fabricmc/fabric/mixin/resource/loader/client/GameOptionsMixin.java
#	fabric-resource-loader-v0/src/client/resources/fabric-resource-loader-v0.client.mixins.json
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModResourcePackCreator.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModResourcePackFactory.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModResourcePackUtil.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/PlaceholderResourcePack.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/DatapackCommandMixin.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/ResourcePackManagerMixin.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/ResourcePackProfileMixin.java
#	fabric-resource-loader-v0/src/main/resources/assets/fabric-resource-loader-v0/lang/zh_cn.json
#	fabric-resource-loader-v0/src/main/resources/assets/fabric-resource-loader-v0/lang/zh_tw.json
#	fabric-resource-loader-v0/src/test/java/net/fabricmc/fabric/test/resource/loader/unit/ModResourcePackUtilTests.java
#	fabric-resource-loader-v0/src/testmod/java/net/fabricmc/fabric/test/resource/loader/BuiltinResourcePackTestMod.java
#	fabric-transfer-api-v1/src/main/resources/fabric-transfer-api-v1.mixins.json
#	gradle.properties

xpple committed:
Update to Minecraft 1.20.6

xpple committed:
Merge remote-tracking branch 'refs/remotes/upstream/1.21' into permissions-api

# Conflicts:
#	deprecated/fabric-convention-tags-v1/src/main/java/net/fabricmc/fabric/impl/tag/convention/ConventionLogWarnings.java
#	fabric-convention-tags-v2/src/datagen/java/net/fabricmc/fabric/impl/tag/convention/datagen/generators/EnglishTagLangGenerator.java
#	fabric-convention-tags-v2/src/generated/resources/assets/fabric-convention-tags-v2/lang/en_us.json
#	fabric-convention-tags-v2/src/main/java/net/fabricmc/fabric/impl/tag/convention/v2/TranslationConventionLogWarnings.java
#	fabric-data-generation-api-v1/src/testmod/java/net/fabricmc/fabric/test/datagen/DataGeneratorTestEntrypoint.java
#	fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/FabricItem.java
#	fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ItemStackMixin.java
#	fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/DefaultItemComponentTest.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModResourcePackCreator.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModResourcePackUtil.java
#	fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/SynchronizeRegistriesTaskMixin.java
#	fabric-resource-loader-v0/src/testmod/java/net/fabricmc/fabric/test/mixin/resource/loader/SynchronizeRegistriesTaskMixin.java
#	fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/impl/screenhandler/Networking.java
#	gradle.properties

xpple committed:
Update to Minecraft 1.21

xpple committed:
Fix Git or something

xpple committed:
Register payload handler
